cpp/src/arrow/compute/exec/data_holder_node.cc

```diff
@@ -0,0 +1,222 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <mutex>
+
+#include "arrow/api.h"
+#include "arrow/compute/api.h"
+
+#include "arrow/compute/memory_resources.h"
+#include "arrow/util/async_generator.h"
+#include "arrow/util/future.h"
+#include "arrow/util/logging.h"
+
+#include "arrow/compute/exec.h"
+#include "arrow/compute/exec/exec_plan.h"
+#include "arrow/compute/exec/options.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/future.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/make_unique.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+
+namespace compute {
+
+class DataHolderManager {
+ public:
+  DataHolderManager(ExecContext* context)
+      : context_(context), gen_(), producer_(gen_.producer()) {}
+
+  Status Push(const std::shared_ptr<RecordBatch>& batch) {
+    static const MemoryLevel all_memory_levels[] = {
+        MemoryLevel::kGPULevel, MemoryLevel::kCPULevel, MemoryLevel::kDiskLevel};
+
+    for (auto id : all_memory_levels) {
+      auto resources = context_->memory_resources();
+
+      auto memory_resource_result = resources->memory_resource(id);
+      if (memory_resource_result.ok()) {
+        auto memory_resource = memory_resource_result.ValueOrDie();
+        auto memory_to_use = memory_resource->memory_used();
+        if (memory_to_use < memory_resource->memory_limit()) {
+          ARROW_ASSIGN_OR_RAISE(auto data_holder, memory_resource->GetDataHolder(batch));
+          this->producer_.Push(std::move(data_holder));
+          break;
+        }
+      }
+    }
+    return Status::OK();
+  }
+  AsyncGenerator<std::shared_ptr<DataHolder>> generator() { return gen_; }
+
+ public:
+  ExecContext* context_;
+  PushGenerator<std::shared_ptr<DataHolder>> gen_;
+  PushGenerator<std::shared_ptr<DataHolder>>::Producer producer_;
+};
+
+class DataHolderNode : public ExecNode {
+ public:
+  DataHolderNode(ExecPlan* plan, NodeVector inputs, std::vector<std::string> input_labels,
+                 std::shared_ptr<Schema> output_schema, int num_outputs)
+      : ExecNode(plan, std::move(inputs), input_labels, std::move(output_schema),
+                 /*num_outputs=*/num_outputs) {
+    executor_ = plan->exec_context()->executor();
+
+    data_holder_manager_ =
+        ::arrow::internal::make_unique<DataHolderManager>(plan->exec_context());
+
+    auto status = task_group_.AddTask([this]() -> Result<Future<>> {
+      ARROW_DCHECK(executor_ != nullptr);
+      return executor_->Submit(this->stop_source_.token(), [this] {
+        auto generator = this->data_holder_manager_->generator();
+        auto iterator = MakeGeneratorIterator(std::move(generator));
```

Comment:
This won't work.  You need to avoid blocking on the CPU thread pool.  This basically takes one of the CPU pool's threads and dedicates it to pulling from the data holder output.  If the data holder fills up then this thread will be a useless thread and since our thread pool is a fixed thread pool this is a bad thing.  Even if the data holder hasn't filled up this thread will spend a lot of time idle if the scanner is slow.

Reply:
Maybe we can use a condition variable that waits until there are things to iterate on. When a generator has something to push it can notify that cv.